src/types/walk.rs

```diff
@@ -5,63 +5,149 @@ extern crate fraction;
 extern crate num_traits;
 extern crate petgraph;
 
-#[derive(Debug, Default, PartialEq, Eq)]
-pub struct RandomWalks<Id> {
-    random_walks_internal: Vec<RandomWalk<Id>>,
+use std::collections::HashMap;
+use std::hash::Hash;
+
+#[derive(Debug, Default)]
+pub struct RandomWalks<Id>
+where
+    Id: Hash + Eq,
+{
+    /// A `HashMap` between the source of the walk and the walk itself.
+    random_walks: HashMap<Id, RandomWalk<Id>>,
 }
 
 impl<Id> RandomWalks<Id>
 where
-    Id: PartialEq,
+    Id: Clone + Eq + Hash,
 {
     pub fn new() -> Self {
         RandomWalks {
-            random_walks_internal: Vec::new(),
+            random_walks: HashMap::new(),
         }
     }
 
     pub fn add_walk(&mut self, walk: RandomWalk<Id>) {
-        self.random_walks_internal.push(walk);
+        self.random_walks
+            .insert(walk.random_walk_source.clone(), walk);
     }
 
     pub fn len(&self) -> usize {
-        self.random_walks_internal.len()
+        self.random_walks.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.random_walks.len() == 0
     }
 
     pub fn count_visits(&self, idx: Id) -> usize {
-        self.random_walks_internal
+        self.random_walks
             .iter()
-            .map(|rw| rw.count_visits(&idx))
+            .map(|(_, rw)| rw.count_visits(&idx))
             .sum()
```

Comment:
Not sure if there is a difference in performance, but I think you could use `values()` instead of `iter()`. See here for more info: https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.values